docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,176 @@
+
+We have some options for allowing users to specify this mode.  It could be a cluster specific flag to the "add" endpoint or a config option setting a cluster wide default.
+
+### Importing with checksums
+If memory/repo size constraints are not a limiting factor it should be straightforward for the cluster-dag importer running on the adding node to keep a running tally of the checksum values and then allocate them to nodes after getting every data shard pinned.  Note this claim is specific to RS as the coding calculations are simple linear combinations of operations and everything commutes, while I wouldn't be surprised if potential future codes also had this property it is something we'll need to check up on once we get serious about pluggability.
+
+If we are in a situation where shards approach the size of ipfs repo or working memory then we can gather inspiration from the report by Plank, specifically the section "Implementation and Performance Details: Checkpointing Systems".  In this section Plank outlines two algorithms for setting checksum values after the data shards are already stored by sending messages over the network.  From my first read-through the broadcast algorithm looks the most promising.  This algorithm would allow cluster to send shards one at a time to the peer holding a zeroed out checksum shard and then perform successive updates to calculate the checksums, rather than requiring that the cluster-dag importer hold the one shard being filled up for pinning alongside the m checksum shards being calculated.
+
+
+### Automatic Recovery
+TODO -- I have thoughts about implementing this in cluster that I need to organize and put down.
+
+# Complications
+
+## Files larger than local repos
+
+### Improve cluster UX for large files
+"How much room do I have in my cluster?" is a very predictable question that users importing huge files will want to know.  We should include cluster cli tools (something like ipfs metrics, or the still-unmade connectivity-graph) that inform users how much space they have available, and the largest recommended file the cluster can ingest.  Perhaps later on we can get even fancier and include predictions of insertion time, and a visualization of how many shards of what size will be added.
+
+Although the ipfs add proxy endpoint is the logical place to launch the adding process we could also provide a user friendly command that takes in the option of a network address to add from.  This way cluster could properly manage the connection with a remote server while buffering and potentially blocking for long periods of time and prevent the user from having to do clever unix things to download large files.  Note I don't have much experience with this kind of thing, so maybe wget etc have really good UX for things like this and this would be overkill.
```

Comment:
the ipfs add proxy endpoint is not the logical point. Cluster may intercept that endpoint and do its thing, but cluster will provide it's own endpoints/methods.

> we could also provide a user friendly command that takes in the option of a network address to add from.  This way cluster could properly manage the connection with a remote server while buffering and potentially blocking for long periods of time and prevent the user from having to do clever unix things to download large files.

You mean like fetching a URL ? Yeah maybe, but let's not worry about this in the first round. It should be an easy thing to do if the rest is working.

Reply:
Got it.  On this topic could you teach me how to do this (e.g. get a buffered, lazily read stream of data from an http endpoint serving, for example, the ubuntu archive into the stdin of ipfs-cluster) with just unix tools and a shell?  It would make the process seem more friendly if I knew how to do it myself.